packages/hardhat-core/src/internal/hardhat-network/provider/utils/makeForkClient.ts

```diff
@@ -1,22 +1,82 @@
-import { BN } from "ethereumjs-util";
+import chalk from "chalk";
+import { BN, toBuffer } from "ethereumjs-util";
 
+import { HARDHAT_NETWORK_NAME } from "../../../constants";
+import { HttpProvider } from "../../../core/providers/http";
 import { JsonRpcClient } from "../../jsonrpc/client";
 import { ForkConfig } from "../node-types";
 
-export async function makeForkClient(forkConfig: ForkConfig) {
-  const forkClient = JsonRpcClient.forUrl(forkConfig.jsonRpcUrl);
+import {
+  FALLBACK_MAX_REORG,
+  getLargestPossibleReorg,
+} from "./reorgs-protection";
+
+export async function makeForkClient(
+  forkConfig: ForkConfig,
+  forkCachePath?: string
+): Promise<{ forkClient: JsonRpcClient; forkBlockNumber: BN }> {
+  const provider = new HttpProvider(
+    forkConfig.jsonRpcUrl,
+    HARDHAT_NETWORK_NAME
+  );
+
+  const networkId = await getNetworkId(provider);
+  const actualMaxReorg = getLargestPossibleReorg(networkId);
+  const maxReorg = actualMaxReorg ?? FALLBACK_MAX_REORG;
+
+  const latestBlock = await getLatestBlockNumber(provider);
+  const lastSafeBlock = latestBlock - maxReorg;
+
   let forkBlockNumber;
   if (forkConfig.blockNumber !== undefined) {
+    if (forkConfig.blockNumber > lastSafeBlock) {
+      const confirmations = latestBlock - forkConfig.blockNumber + 1;
+      const requiredConfirmations = maxReorg + 1;
```

Comment:
I don't understand these two `+ 1`.

Reply:
Just a matter of what "confirmation" means in this context. When your transaction gets mined, we say that it has 1 confirmation, but it's actually in the latest block. 

So if you are forking from the latest block, `latestBlock - forkConfig.blockNumber === 0`, and yet, we say it has 1 confirmation. So `+ 1`.